<a href="https://colab.research.google.com/github/MarioROT/IHLT-MAI/blob/main/Session3_Pruebas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab session 3 (Morphology) - IHLT

**Students:**
- Mario Rosas
- Alam Lopez

**Lab Professor:** Salvador Medina Herrera

## Paraphrases Template

In [1]:
%%shell
git clone https://github.com/mariorot/IHLT-MAI.git
cd 'IHLT-MAI'
mv 'complementary_material' /content/
mv scripts /content/

fatal: destination path 'IHLT-MAI' already exists and is not an empty directory.
mv: cannot stat 'complementary_material': No such file or directory
mv: cannot stat 'scripts': No such file or directory


CalledProcessError: ignored

In [2]:
from scripts.compute_metrics import ComputeMetrics
from scripts.text_preprocessing import TextPreprocessing
import pandas as pd
from scipy.stats import pearsonr
import nltk
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
nltk.download('omw-1.4')
import spacy
nlp = spacy.load("en_core_web_sm")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/

In [3]:
dt = pd.read_csv('complementary_material/test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)
dt['gs'] = pd.read_csv('complementary_material/test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)

## TODO
1. Read all pairs of sentences of the SMTeuroparl files of test set within the evaluation framework of the project. Compute their similarities by considering *lemmas* and *Jaccard distance*. <br>

2. Compare the results with those in session 2 (document structure) in which words were considered. : <br>

3. Compare the results with gold standard by giving the pearson correlation between them. <br>

4. Questions (justify the answers): <br>

      Which is better: words or lemmas?

      Do you think that could perform better for any pair of texts?
```
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]
```

The code for the tokenization is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/text_preprocessing.py

The code for jaccard distance computation is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/compute_metrics.py

In [4]:
tp = TextPreprocessing()
# --Tokenization
# NLTK tokenization
dt[2] = tp.tokenize_data(list(dt[0]),'nltk')
dt[3] = tp.tokenize_data(list(dt[1]),'nltk')
# spaCy tokenization
dt[4] = tp.tokenize_data(list(dt[0]),'spacy')
dt[5] = tp.tokenize_data(list(dt[1]),'spacy')

# --Lemmatization
# NLTK lemmatization
dt[6] = tp.lemmatize_data(list(dt[0]),'nltk', False)
dt[7] = tp.lemmatize_data(list(dt[1]),'nltk', False)
# spaCy lemmatization
dt[8] = tp.lemmatize_data(list(dt[0]),'spacy')
dt[9] = tp.lemmatize_data(list(dt[1]),'spacy')

# --Lemmatization with cleaned data
# NLTK lemmatization
dt[10] = tp.lemmatize_data(tp.clean_data(list(dt[0])),'nltk', False)
dt[11] = tp.lemmatize_data(tp.clean_data(list(dt[1])),'nltk', False)
# spaCy lemmatization
dt[12] = tp.lemmatize_data(tp.clean_data(list(dt[0])),'spacy')
dt[13] = tp.lemmatize_data(tp.clean_data(list(dt[1])),'spacy')

In [5]:
dt

0  \
0    The leaders have now been given a new chance a...   
1    Amendment No 7 proposes certain changes in the...   
2    Let me remind you that our allies include ferv...   
3          The vote will take place today at 5.30 p.m.   
4    The fishermen are inactive, tired and disappoi...   
..                                                 ...   
454  It is our job to continue to support Latvia wi...   
455        The vote will take place today at 5.30 p.m.   
456  Neither was there a qualified majority within ...   
457  Let me remind you that our allies include ferv...   
458  We often pontificate here about being the repr...   

                                                     1     gs  \
0    The leaders benefit aujourd' hui of a new luck...  4.500   
1    Amendment No 7 is proposing certain changes in...  5.000   
2    I would like to remind you that among our alli...  4.250   
3                   The vote will take place at 5.30pm  4.500   
4    The fishermen are inactive, tired and disappoi...  5.000   
..                                                 ...    ...   
454  It is of our duty of continue to support the c...  5.000   
455                   Vote will take place at 17 h 30.  4.750   
456  There was no qualified majority in this Parlia...  5.000   
457  I hold you recall that our allies, there are e...  4.000   
458  We often take pride here to represent the citi...  3.833   

                                                     2  \
0    [The, leaders, have, now, been, given, a, new,...   
1    [Amendment, No, 7, proposes, certain, changes,...   
2    [Let, me, remind, you, that, our, allies, incl...   
3    [The, vote, will, take, place, today, at, 5.30...   
4    [The, fishermen, are, inactive, ,, tired, and,...   
..                                                 ...   
454  [It, is, our, job, to, continue, to, support, ...   
455  [The, vote, will, take, place, today, at, 5.30...   
456  [Neither, was, there, a, qualified, majority, ...   
457  [Let, me, remind, you, that, our, allies, incl...   
458  [We, often, pontificate, here, about, being, t...   

                                                     3  \
0    [The, leaders, benefit, aujourd, ', hui, of, a...   
1    [Amendment, No, 7, is, proposing, certain, cha...   
2    [I, would, like, to, remind, you, that, among,...   
3           [The, vote, will, take, place, at, 5.30pm]   
4    [The, fishermen, are, inactive, ,, tired, and,...   
..                                                 ...   
454  [It, is, of, our, duty, of, continue, to, supp...   
455        [Vote, will, take, place, at, 17, h, 30, .]   
456  [There, was, no, qualified, majority, in, this...   
457  [I, hold, you, recall, that, our, allies, ,, t...   
458  [We, often, take, pride, here, to, represent, ...   

                                                     4  \
0    [The, leaders, have, now, been, given, a, new,...   
1    [Amendment, No, 7, proposes, certain, changes,...   
2    [Let, me, remind, you, that, our, allies, incl...   
3    [The, vote, will, take, place, today, at, 5.30...   
4    [The, fishermen, are, inactive, ,, tired, and,...   
..                                                 ...   
454  [It, is, our, job, to, continue, to, support, ...   
455  [The, vote, will, take, place, today, at, 5.30...   
456  [Neither, was, there, a, qualified, majority, ...   
457  [Let, me, remind, you, that, our, allies, incl...   
458  [We, often, pontificate, here, about, being, t...   

                                                     5  \
0    [The, leaders, benefit, aujourd, ', hui, of, a...   
1    [Amendment, No, 7, is, proposing, certain, cha...   
2    [I, would, like, to, remind, you, that, among,...   
3           [The, vote, will, take, place, at, 5.30pm]   
4    [The, fishermen, are, inactive, ,, tired, and,...   
..                                                 ...   
454  [It, is, of, our, duty, of, continue, to, supp...   
455        [Vote, wi

In [6]:
cm_nltk = ComputeMetrics(dt[[2,3]].to_numpy(), ['jaccard'], 1)
dt['jaccard_nltk'] = cm_nltk.do()['jaccard']

cm_spacy = ComputeMetrics(dt[[4,5]].to_numpy(), ['jaccard'], 1)
dt['jaccard_spacy'] = cm_spacy.do()['jaccard']

cm_spacy = ComputeMetrics(dt[[6,7]].to_numpy(), ['jaccard'], 1)
dt['jaccard_lemmas_nltk'] = cm_spacy.do()['jaccard']

cm_spacy = ComputeMetrics(dt[[8,9]].to_numpy(), ['jaccard'], 1)
dt['jaccard_lemmas_spacy'] = cm_spacy.do()['jaccard']

cm_spacy = ComputeMetrics(dt[[10,11]].to_numpy(), ['jaccard'], 1)
dt['jaccard_clean_lemmas_nltk'] = cm_spacy.do()['jaccard']

cm_spacy = ComputeMetrics(dt[[12,13]].to_numpy(), ['jaccard'], 1)
dt['jaccard_clean_lemmas_spacy'] = cm_spacy.do()['jaccard']

Computing jaccard...
Computing jaccard...
Computing jaccard...
Computing jaccard...
Computing jaccard...
Computing jaccard...


In [7]:
dt[['gs','jaccard_nltk','jaccard_spacy','jaccard_lemmas_nltk','jaccard_lemmas_spacy', 'jaccars_clean_lemmas_s']]

gs  jaccard_nltk  jaccard_spacy  jaccard_lemmas_nltk  \
0    4.500      0.653846       0.653846             0.653846   
1    5.000      0.214286       0.214286             0.076923   
2    4.250      0.608696       0.608696             0.608696   
3    4.500      0.454545       0.400000             0.454545   
4    5.000      0.000000       0.000000             0.000000   
..     ...           ...            ...                  ...   
454  5.000      0.450000       0.450000             0.450000   
455  4.750      0.642857       0.714286             0.642857   
456  5.000      0.600000       0.600000             0.600000   
457  4.000      0.681818       0.681818             0.681818   
458  3.833      0.500000       0.500000             0.500000   

     jaccard_lemmas_spacy  
0                0.600000  
1                0.076923  
2                0.545455  
3                0.400000  
4                0.000000  
..                    ...  
454              0.450000  
455              0.615385  
456              0.526316  
457              0.619048  
458              0.500000  

[459 rows x 5 columns]

In [14]:
p_nltk=pearsonr(dt['gs'], 1-dt['jaccard_nltk'])[0] # Calculating the pearson correlation between GS results and 1-Jaccard calculated dats
p_spacy=pearsonr(dt['gs'], 1-dt['jaccard_spacy'])[0]
p_l_nltk=pearsonr(dt['gs'], 1-dt['jaccard_lemmas_nltk'])[0]
p_l_spacy=pearsonr(dt['gs'], 1-dt['jaccard_lemmas_spacy'])[0]
p_l_nltk=pearsonr(dt['gs'], 1-dt['jaccard_lemmas_nltk'])[0]
p_l_spacy=pearsonr(dt['gs'], 1-dt['jaccard_lemmas_spacy'])[0]
print(f'NLTK {p_nltk} --- spaCy {p_spacy} --- NLTK Lemmas {p_l_nltk} --- spaCy Lemmas {p_l_spacy}')

NLTK 0.45049771693186846 --- spaCy 0.4608504322255335 --- NLTK Lemmas 0.4492874506144061 --- spaCy Lemmas nan


/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:4781: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(stats.ConstantInputWarning(msg))


In [ ]:
tp.lemmatize_data([words], 'nltk')

[['woman', 'play', 'with', 'small', 'child', 'happily']]

In [ ]:
tp.lemmatize_data([' '.join(words)], 'spacy')

[['woman', 'play', 'with', 'small', 'child', 'happily']]

In [ ]:
' '.join(words)

'women played with small children happily'

# Conclusion
The use of different tokenizers (NLTK and spaCy) can impact the results when comparing sentence pairs, as they may perform differently depending on the input text data. Jaccard similarity proved to be an effective metric for comparing the similarity of sentence pairs after word tokenization. With the inversion of this metric we got a distance to measure the dissimilarity between sets of tokens, making it suitable for the task we were trying to solve. The Pearson correlation coefficient provided us a quantitative measure of how well our results align with the gold-standard, i.e. effectiveness.

The results indicate that spaCy slightly outperformed NLTK in the task of sentence pair comparison using Jaccard distance.The Pearson correlation coefficient values obtained for **NLTK (0.4505)** and **spaCy (0.4609)** suggest a moderate positive correlation between the computed distances and the benchmark (gold-standard) distances.

Our best result could be possibly improved by experimenting with different tokenization strategies or exploring other NLP techniques to enhance the performance.


